# Exercise 1

In this activity, you will work with the New York City Airbnb Open Data—a popular, messy, and highly realistic dataset used by data analysts and data scientists around the world. The dataset includes thousands of Airbnb listings along with details about hosts, locations, prices, reviews, and more.

<img src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcR8Pmq3Gv7y7z_2Xun-OqMIk43kd5u0TXEatw&s"/>


To guide your work, you will follow the 6-Step Data Wrangling Process, a professional workflow used in industry:

- Discovering. Become familiar with the data by exploring its structure and observing patterns.

- Structuring. Fix issues related to format, column types, indices, and duplicated records.

- Cleaning. Handle missing values, outliers, and inconsistencies so the dataset becomes trustworthy.

- Enriching. Add new features, aggregate information, and enhance the dataset for deeper insights.

- Validating. Apply rule-based checks to ensure the data meets quality standards.

- Publishing. Export and prepare the cleaned dataset for downstream analysis or reporting.
Throughout this exercise, you will answer questions and fill in code cells based on these six steps. You will mimic what professional data scientists do when preparing data for business intelligence dashboards, machine learning models, or exploratory analysis.

In [24]:
import kagglehub
import os
import pandas as pd
import numpy as np

# Download latest version
path = kagglehub.dataset_download("dgomonov/new-york-city-airbnb-open-data")

print("Path to dataset files:", path)

if os.path.isdir(path):
  print(True)

contents = os.listdir(path)
contents

mydataset = path + "/" + contents[0]
mydataset


df = pd.read_csv(mydataset)

Using Colab cache for faster access to the 'new-york-city-airbnb-open-data' dataset.
Path to dataset files: /kaggle/input/new-york-city-airbnb-open-data
True


## A. Discovering


Display the first 10 rows

In [25]:
df.head(10)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   
5  5099         Large Cozy 1 BR Apartment In Midtown East     7322   
6  5121                                   BlissArtsSpace!     7356   
7  5178                  Large Furnished Room Near B'way      8967   
8  5203                Cozy Clean Guest Room - Family Apt     7490   
9  5238                Cute & Cozy Lower East Side 1 bdrm     7549   

     host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0         John            Brooklyn          Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan             Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan              Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan         East Harlem  40.79851  -73.94399   
5        Chris           Manhattan         Murray Hill  40.74767  -73.97500   
6        Garon            Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   
7     Shunichi           Manhattan      Hell's Kitchen  40.76489  -73.98493   
8    MaryEllen           Manhattan     Upper West Side  40.80178  -73.96723   
9          Ben           Manhattan           Chinatown  40.71344  -73.99037   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   
5  Entire home/apt    200               3                 74  2019-06-22   
6     Private room     60              45                 49  2017-10-05   
7     Private room     79               2                430  2019-06-24   
8     Private room     79               2                118  2017-07-21   
9  Entire home/apt    150               1                160  2019-06-09   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0  
5               0.59                               1               129  
6               0.40                               1                 0  
7               3.47                               1               220  
8               0.99                               1                 0  
9               1.33                               4               188

Display the column names:

In [26]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

Display the dataframe shape:

In [27]:
df.shape

(48895, 16)

Identify the numerical features:

In [28]:
df.select_dtypes(include=np.number).columns


Index(['id', 'host_id', 'latitude', 'longitude', 'price', 'minimum_nights',
       'number_of_reviews', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365'],
      dtype='object')

Identify the categorical features

In [29]:
df.select_dtypes(include='object').columns

Index(['name', 'host_name', 'neighbourhood_group', 'neighbourhood',
       'room_type', 'last_review'],
      dtype='object')

What are the top 5 questions this dataset might help answer?

In [32]:
questions = [
    "Which neighborhood has the most Airbnb listings?",
    "Do highly available listings also get more reviews?",
    "How does the price change based on the type of room being rented?",
    "Do hosts with many listings charge more or less than hosts with just a few?",
    "What is the most common required length of stay, and does a longer stay lead to fewer reviews?"
]

for i, q in enumerate(questions, 1):
    print(f"{i}. {q}")

1. Which neighborhood has the most Airbnb listings?
2. Do highly available listings also get more reviews?
3. How does the price change based on the type of room being rented?
4. Do hosts with many listings charge more or less than hosts with just a few?
5. What is the most common required length of stay, and does a longer stay lead to fewer reviews?


## B. Structuring

Check if the dataset has proper column types

In [65]:
df.info

<bound method DataFrame.info of              id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaT               0.00                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaT               0.00                               2   
48891         NaT               0.00                               2   
48892         NaT               0.00                               1   
48893         NaT               0.00                               6   
48894         NaT               0.00                               1   

       availability_365  price_per_minimum_night  review_rate_per_month  \
0                   365               149.000000                   1.89   
1                   355               225.000000                  17.10   
2                   365                50.000000                   0.00   
3   

Check if the dataset has index issues

In [35]:
print(df.index)
print(f"Is index unique: {df.index.is_unique}")
print(f"Is index monotonic: {df.index.is_monotonic_increasing}")

RangeIndex(start=0, stop=48895, step=1)
Is index unique: True
Is index monotonic: True


Check if the dataset has duplicated rows

In [36]:
print("Total number of exact duplicated rows:", df.duplicated().sum())

df[df.duplicated(keep=False)].sort_values(by='id')


Total number of exact duplicated rows: 0


Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

Convert columns to correct types (e.g., dates → datetime).

In [37]:
df['last_review'] = pd.to_datetime(df['last_review'])
print(df['last_review'].dtype)

datetime64[ns]


Identify at least one structural issue and describe how to fix it.

In [43]:
categorical_cols = df.select_dtypes(include='object').columns

for col in categorical_cols:
    df[col] = df[col].astype('category')

df[categorical_cols].dtypes

Series([], dtype: object)

## C. Cleaning



Check for missing values

In [44]:
df.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

Check for outliers in `minimum_nights`

In [45]:
df['minimum_nights'].describe()

count    48895.000000
mean         7.029962
std         20.510550
min          1.000000
25%          1.000000
50%          3.000000
75%          5.000000
max       1250.000000
Name: minimum_nights, dtype: float64

Check for invalid values (e.g., price ≤ 0)

In [47]:
df[df['price'] <= 0].shape[0]


11

Apply a missing-value strategy (drop, fill, or flag)


In [50]:
df['reviews_per_month'] = df['reviews_per_month'].fillna(0.0)

if 'Unknown' not in df['host_name'].cat.categories:
    df['host_name'] = df['host_name'].cat.add_categories('Unknown')

df['host_name'] = df['host_name'].fillna('Unknown')

if 'Unknown Listing Name' not in df['name'].cat.categories:
    df['name'] = df['name'].cat.add_categories('Unknown Listing Name')
df['name'] = df['name'].fillna('Unknown Listing Name')

print("Missing value imputation complete. Check the new null counts.")
print(df.isnull().sum())

Missing value imputation complete. Check the new null counts.
id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64


Apply an outlier-handling strategy

In [ ]:
# put your answer here

## D. Enriching

Create at least 3 new variables:
- price_per_minimum_night
- review_rate_per_month (combine ratings + frequency)
- is_expensive (boolean flag based on price threshold)


In [51]:
df['price_per_minimum_night'] = df['price'] / df['minimum_nights']

df['review_rate_per_month'] = df['number_of_reviews'] * df['reviews_per_month']

price_75th_percentile = df['price'].quantile(0.75)
df['is_expensive'] = (df['price'] > price_75th_percentile)

print(f"Price threshold for 'is_expensive' (75th Percentile): ${price_75th_percentile:.2f}")
print("New features created. Displaying head with new columns:")
df[['price', 'minimum_nights', 'price_per_minimum_night', 'number_of_reviews', 'reviews_per_month', 'review_rate_per_month', 'is_expensive']].head()


Price threshold for 'is_expensive' (75th Percentile): $175.00
New features created. Displaying head with new columns:


price  minimum_nights  price_per_minimum_night  number_of_reviews  \
0    149               1                    149.0                  9   
1    225               1                    225.0                 45   
2    150               3                     50.0                  0   
3     89               1                     89.0                270   
4     80              10                      8.0                  9   

   reviews_per_month  review_rate_per_month  is_expensive  
0               0.21                   1.89         False  
1               0.38                  17.10          True  
2               0.00                   0.00         False  
3               4.64                1252.80         False  
4               0.10                   0.90         False

Create 1 aggregated summary table

(e.g. average price per neighborhood, number of hosts per borough)

In [58]:
neighbourhood_summary = df.groupby('neighbourhood_group').agg(
    average_price=('price', 'mean'),
    median_price=('price', 'median'),
    total_listings=('id', 'count'),
    total_unique_hosts=('host_id', 'nunique')
).reset_index()

print("Aggregated Summary Table (Price and Host Count per Borough):")
neighbourhood_summary

Aggregated Summary Table (Price and Host Count per Borough):


/tmp/ipython-input-2497029315.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  neighbourhood_summary = df.groupby('neighbourhood_group').agg(


neighbourhood_group  average_price  median_price  total_listings  \
0               Bronx      87.577064          65.0            1090   
1            Brooklyn     124.438915          90.0           20095   
2           Manhattan     196.884903         150.0           21660   
3              Queens      99.517649          75.0            5666   
4       Staten Island     114.812332          75.0             373   

   total_unique_hosts  
0                 789  
1               15965  
2               16577  
3                3983  
4                 256

## E. Validating

Create at least 3 validation checks
- Price must be positive.
- Host ID should not be null.
- Minimum nights must be ≥ 1.
- Generate a validation report:
- Count of rows violating each rule.

Decide if invalid rows (if any) should be removed or corrected?

In [64]:
price_violation = df['price'] <= 0
host_id_violation = df['host_id'].isnull()
min_nights_violation = df['minimum_nights'] < 1

# Generate the Validation Report
validation_report = pd.DataFrame({
    'Rule Violated': [
        'Price must be positive',
        'Host ID should not be null',
        'Minimum nights must be ≥ 1'
    ],
    'Violating Rows Count': [
        price_violation.sum(),
        host_id_violation.sum(),
        min_nights_violation.sum()
    ]
})

print("VALIDATION REPORT")
print(validation_report)

VALIDATION REPORT
                Rule Violated  Violating Rows Count
0      Price must be positive                     0
1  Host ID should not be null                     0
2  Minimum nights must be ≥ 1                     0


## F. Publishing

Select final columns for publication.



In [60]:
final_columns = [
    'id', 'name', 'host_id', 'host_name', 'neighbourhood_group', 'neighbourhood',
    'latitude', 'longitude', 'room_type', 'price', 'minimum_nights',
    'number_of_reviews', 'last_review', 'reviews_per_month',
    'calculated_host_listings_count', 'availability_365',
    'price_per_minimum_night', 'review_rate_per_month', 'is_expensive'
]

df_final = df[final_columns].copy()

print("Final dataset prepared with 19 columns, including 3 new enriched features.")
print(df_final.head(3))


Final dataset prepared with 19 columns, including 3 new enriched features.
     id                                 name  host_id  host_name  \
0  2539   Clean & quiet apt home by the park     2787       John   
1  2595                Skylit Midtown Castle     2845   Jennifer   
2  3647  THE VILLAGE OF HARLEM....NEW YORK !     4632  Elisabeth   

  neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0            Brooklyn    Kensington  40.64749  -73.97237     Private room   
1           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt   
2           Manhattan        Harlem  40.80902  -73.94190     Private room   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0    149               1                  9  2018-10-19               0.21   
1    225               1                 45  2019-05-21               0.38   
2    150               3                  0         NaT               0.00   

   calculated_host_listings_co

Export cleaned dataset as "cleaned_airbnb.csv"

In [63]:
df_final.to_csv("cleaned_airbnb.csv", index=False)

print("SUCCESS: 'cleaned_airbnb.csv' exported.")

SUCCESS: 'cleaned_airbnb.csv' exported.


Export summary reports (e.g., validation or aggregations) as CSV.

In [62]:
neighbourhood_summary.to_csv("neighbourhood_summary_report.csv", index=False)

validation_report.to_csv("validation_report.csv", index=False)

print("SUCCESS: Summary reports exported.")

SUCCESS: Summary reports exported.


# Task
Remove rows from the DataFrame where 'price' is less than or equal to 0, and then cap the 'minimum_nights' values at the 99th percentile. Finally, display descriptive statistics for `minimum_nights` and the number of rows removed due to invalid prices.

## remove_invalid_prices

### Subtask:
Remove rows from the DataFrame where the 'price' column is less than or equal to 0, as these are considered invalid data points.


**Reasoning**:
First, I will count the number of rows where the 'price' column is less than or equal to 0 to quantify the data to be removed. This step aligns with the instruction to store the count in a variable named `initial_rows_removed`.



In [52]:
initial_rows_removed = df[df['price'] <= 0].shape[0]
print(f"Number of rows with invalid price (<= 0) to be removed: {initial_rows_removed}")

Number of rows with invalid price (<= 0) to be removed: 11


**Reasoning**:
Now that the number of rows with invalid prices has been counted, the next step is to filter the DataFrame to remove these rows, keeping only those where the 'price' is greater than 0, and then update the DataFrame.



In [53]:
df = df[df['price'] > 0]
print(f"DataFrame shape after removing invalid prices: {df.shape}")

DataFrame shape after removing invalid prices: (48884, 19)


# Task
Identify the 99th percentile for the 'minimum_nights' column and replace any values above this threshold with the 99th percentile value to handle extreme outliers and normalize the distribution. After this, display descriptive statistics for `minimum_nights` and the number of rows removed due to invalid prices.

## cap_minimum_nights_outliers

### Subtask:
Identify the 99th percentile for the 'minimum_nights' column and replace any values above this threshold with the 99th percentile value to handle extreme outliers and normalize the distribution. After this, display descriptive statistics for `minimum_nights` and the number of rows removed due to invalid prices.


**Reasoning**:
To cap the outliers in 'minimum_nights', I will first calculate its 99th percentile, which serves as the upper bound for acceptable values, and then print it for verification.



In [54]:
percentile_99 = df['minimum_nights'].quantile(0.99)
print(f"99th percentile for minimum_nights: {percentile_99}")

99th percentile for minimum_nights: 45.0


**Reasoning**:
Now that the 99th percentile has been calculated, I will replace all 'minimum_nights' values exceeding this threshold with the 99th percentile value to cap outliers.



In [55]:
df.loc[df['minimum_nights'] > percentile_99, 'minimum_nights'] = percentile_99
print(f"'minimum_nights' values capped at {percentile_99}.")

'minimum_nights' values capped at 45.0.


**Reasoning**:
To confirm the effect of capping and understand the distribution of the 'minimum_nights' column after outlier handling, I will display its descriptive statistics.



In [56]:
print("Descriptive statistics for 'minimum_nights' after capping:")
print(df['minimum_nights'].describe())

Descriptive statistics for 'minimum_nights' after capping:
count    48884.000000
mean         6.116480
std          9.244575
min          1.000000
25%          1.000000
50%          3.000000
75%          5.000000
max         45.000000
Name: minimum_nights, dtype: float64


**Reasoning**:
To complete the subtask, I will now print the number of rows that were initially removed due to invalid prices, using the `initial_rows_removed` variable.



In [57]:
print(f"Number of rows removed due to invalid prices: {initial_rows_removed}")

Number of rows removed due to invalid prices: 11


## final_task

### Subtask:
Summarize the outlier handling strategies applied and their impact on the dataset.


## Summary:

### Q&A
1.  **What is the 99th percentile for the 'minimum_nights' column?**
    The 99th percentile for the 'minimum_nights' column is 45.0.
2.  **What are the descriptive statistics for 'minimum_nights' after capping?**
    After capping, the 'minimum_nights' column has a count of 48884, a mean of 6.116, a standard deviation of 9.244, a minimum of 1.0, a 25th percentile of 1.0, a median (50th percentile) of 3.0, a 75th percentile of 5.0, and a maximum value of 45.0.
3.  **How many rows were removed due to invalid prices?**
    11 rows were removed due to invalid prices in a previous step.

### Data Analysis Key Findings
*   The 99th percentile for the `minimum_nights` column was identified as 45.0.
*   All values in `minimum_nights` exceeding 45.0 were successfully capped at this threshold, effectively handling extreme outliers.
*   After capping, the maximum value for `minimum_nights` is now 45.0, down from its original uncapped value, indicating the successful application of the outlier treatment.
*   The mean of `minimum_nights` after capping is 6.116, with a standard deviation of 9.244, suggesting a less skewed distribution compared to the uncapped state.
*   A total of 11 rows had been previously removed from the dataset due to invalid prices.

### Insights or Next Steps
*   The capping of `minimum_nights` at the 99th percentile has effectively mitigated the impact of extreme outliers, leading to a more normalized and potentially more robust distribution for subsequent analysis or model training.
*   Consider further exploring the distribution of `minimum_nights` (e.g., through histograms or box plots) to evaluate the overall impact of capping on its shape and to identify any remaining skewness that might require alternative transformation techniques.
